### PR TITLE
Add method UserSession#keepAuthorizedComponents()

### PR DESCRIPTION
**Pull request by simonbrandhof**
to be able to fix authorization performance hotspot

### CODE REVIEW

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java [hunk 1]
```diff
@@ -135,6 +135,17 @@ public int countUsersWithGlobalPermissionExcludingUserPermission(DbSession dbSes
       });
   }
 
+  public Set<String> keepAuthorizedProjectUuids(DbSession dbSession, Collection<String> projectUuids, @Nullable Integer userId, String permission) {
+    return executeLargeInputsIntoSet(
+      projectUuids,
+      partition -> {
+        if (userId == null) {
+          return mapper(dbSession).keepAuthorizedProjectUuidsForAnonymous(permission, projectUuids);
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: 💣  Change `projectUuids` to `partition`!

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java [hunk 2]
```diff
@@ -135,6 +135,17 @@ public int countUsersWithGlobalPermissionExcludingUserPermission(DbSession dbSes
...
+        }
+        return mapper(dbSession).keepAuthorizedProjectUuidsForUser(userId, permission, projectUuids);
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: 💣  Change `projectUuids` to `partition`!

File: server/sonar-db-dao/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml [hunk 3]
```diff
@@ -219,17 +219,72 @@
...
-      projects p
+    projects p
     where
-      <foreach collection="componentIds" open="(" close=")" item="element" index="index" separator=" or ">
-        p.id=#{element,jdbcType=BIGINT}
-      </foreach>
-      and p.private = ${_false}
-      and #{role,jdbcType=VARCHAR} in ('user','codeviewer')
+    <foreach collection="componentIds" open="(" close=")" item="element" index="index" separator=" or ">
+      p.id=#{element,jdbcType=BIGINT}
```
Comment (bartfastiel) [inline server/sonar-db-dao/src/main/resources/org/sonar/db/permission/AuthorizationMapper.xml]: Wouldn't `IN` be better than `=`? (Although that change should probably not be part of this commit)

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 4]
```diff
@@ -181,6 +182,20 @@ protected boolean hasProjectUuidPermission(String permission, String projectUuid
   }
 
   @Override
+  protected List<ComponentDto> doKeepAuthorizedComponents(String permission, Collection<ComponentDto> components) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Set<String> projectUuids = components.stream()
+        .map(ComponentDto::projectUuid)
+        .collect(MoreCollectors.toSet());
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: `toSet(components.size())` :stuck_out_tongue:

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java [hunk 5]
```diff
@@ -181,6 +182,20 @@ protected boolean hasProjectUuidPermission(String permission, String projectUuid
...
+      Set<String> authorizedProjectUuids = dbClient.authorizationDao().keepAuthorizedProjectUuids(dbSession, projectUuids, getUserId(), permission);
+
+      return components.stream()
+        .filter(c -> authorizedProjectUuids.contains(c.projectUuid()))
+        .collect(MoreCollectors.toList());
```
Comment (bartfastiel) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: We could think of using `toList(authorizedProjectUuids.size())`

File: server/sonar-server/src/test/java/org/sonar/server/tester/MockUserSession.java [hunk 6]
```diff
@@ -23,6 +23,7 @@
 import java.util.Collection;
 import java.util.List;
 import java.util.Objects;
+import org.sonar.db.component.ComponentDto;
```
Comment (bartfastiel) [inline server/sonar-server/src/test/java/org/sonar/server/tester/MockUserSession.java]: If this is the only change in this file, this import is most likely not needed.

File: server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java (hunk 4 above)
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/user/ServerUserSession.java]: are you sure ? There can be much more components than projects.

File: server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java (hunk 1 above)
Comment (simonbrandhof) [inline server/sonar-db-dao/src/main/java/org/sonar/db/permission/AuthorizationDao.java]: Incredible, that means that there is the same bug in keepAuthorizedProjectIds() !
